<a href="https://colab.research.google.com/github/JoshScherer/Emergency-Response/blob/josh-dev/FiguringOutSpark.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"

In [ ]:
#install spark. we are using the one that uses hadoop as the underlying scheduler.
!wget -q https://downloads.apache.org/spark/spark-3.2.4/spark-3.2.4-bin-hadoop3.2.tgz
!tar xf  spark-3.2.4-bin-hadoop3.2.tgz
!ls -l

#Provides findspark.init() to make pyspark importable as a regular library.
os.environ["SPARK_HOME"] = "spark-3.2.4-bin-hadoop3.2"
os.environ['PYSPARK_SUBMIT_ARGS'] = '--packages org.apache.spark:spark-streaming-kafka-0-8_2.11:2.4.7 pyspark-shell'

total 294140
drwxr-xr-x  1 root root      4096 Apr 27 13:35 sample_data
drwxr-xr-x 13 1000 1000      4096 Apr  9 21:17 spark-3.2.4-bin-hadoop3.2
-rw-r--r--  1 root root 301183180 Apr  9 21:35 spark-3.2.4-bin-hadoop3.2.tgz


In [ ]:
!unzip nashville-tweets-2019-01-28.zip

Archive:  nashville-tweets-2019-01-28.zip
  inflating: nashville-tweets-2019-01-28  


In [ ]:
!head -3 nashville-tweets-2019-01-28.zip

�J��Zn������}�YI_4+Z���
�zf�صj�n�oDx�^G]8�s����m9�u����a��s��K�V��yۯ|�w^f��&�d�LX#9p���g�G}��O�Ԕ�q<�����>s���g���Vh�/7�p�z�I{jd�7à?>���Wv�l�
:58�)L3���(��������:���>~�����;��VՅ����_�eh�����_����9i!q/�������e~}[�:��Tjv�g.�ס�+��


In [ ]:
!pip install -q findspark pyspark
import findspark
findspark.init()

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.8/310.8 MB 3.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


In [ ]:
# LOCAL
%%file local_1_count.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark import SparkContext, SparkConf
import json
from operator import add
#import other things as required

#create spark context. This is very important. Do this similarly for the other parts
# Note to read a file directly from s3 into an rdd you may have to do something like this

def checkjson(entry):
  try:
    json.loads(entry)
    #if load succeeded. We use correct as the key
    return "correct", 1
  except:
    #there was an error in loading. We use incorrect as the key
    return "incorrect", 1


if __name__ == '__main__':
  # replace this line with the s3 pass when testing over EMR
  conf = SparkConf().setAppName('local_1_count').set('spark.hadoop.validateOutputSpecs', False)
  sc = SparkContext(conf=conf).getOrCreate()

  try:
    #@todo: fix the path as required
    tweets=sc.textFile('/content/nashville-tweets-2019-01-28')
    # review the page rank example for how to use the map operation
    # review word count for reduce and add
    # see how we use map to parse each row
    counts = tweets.map(lambda tweet: checkjson(tweet)).reduceByKey(add)
    print(counts.take(2))

    # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
    output = counts.sortBy(lambda a: a[1], ascending=False)

    # @todo: the s3 version will have to save it to correct s3 path
    output.repartition(1).saveAsTextFile("local_1_count.out")
    

  finally:
    # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
    #finally is used to make sure the context is stopped even with errors
    sc.stop()
  

 
  
  pass

Writing local_1_count.py


In [ ]:
!spark-submit --packages org.apache.spark:spark-streaming-kafka-0-8_2.11:2.4.7 local_1_count.py


:: loading settings :: url = jar:file:/content/spark-3.2.4-bin-hadoop3.2/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /root/.ivy2/cache
The jars for the packages stored in: /root/.ivy2/jars
org.apache.spark#spark-streaming-kafka-0-8_2.11 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-dce22bc7-e519-41a8-b87c-92cc415d9e8c;1.0
	confs: [default]
	found org.apache.spark#spark-streaming-kafka-0-8_2.11;2.4.7 in central
	found org.apache.kafka#kafka_2.11;0.8.2.1 in central
	found org.scala-lang.modules#scala-xml_2.11;1.0.2 in central
	found com.yammer.metrics#metrics-core;2.2.0 in central
	found org.slf4j#slf4j-api;1.7.16 in central
	found org.scala-lang.modules#scala-parser-combinators_2.11;1.1.0 in central
	found com.101tec#zkclient;0.3 in central
	found log4j#log4j;1.2.17 in central
	found org.apache.kafka#kafka-clients;0.8.2.1 in central
	found net.jpountz.lz4#lz4;1.2.0 in central
	found org.xerial.snapp

In [11]:
credentials = {
    'region_name': 'us-east-1',
    'aws_access_key_id': 'ASIAYELUFHJM2IAO5LOI',
    'aws_secret_access_key': 'Kicby5su+q8Io/ZdX25RI/hfAdqsYfGlO4k/maUg',
    'aws_session_token': 'FwoGZXIvYXdzEI3//////////wEaDF+YaQijw81lcUGJtyLMAcvxVUB1kRKaxxo1VJbOjaHFR959YX6uaibZ4CbJSBKIOGk8x8PsDIZ5GXCTtJKg5nFpu2YNi1sSGFD68tU1zoq6SkKyTvXfIAm9VE2X9UDg2Tr7a1JgCLLpzRnXYiMslGh1KB/M8olhJrcv5ITXu93fPYzskdebAakk347vxVmfKOsBaOG6iAuCfvKr4dkVu8gTFsWRbmxK/Z+ZQWHkAiB4zcoMx8FdW4Lfi9jB8e2w2nBd23rVM/Ioz6Uc+J5kCG0/9kbzKLiDvHESaSjmubeiBjItWG+zzDJ0zoEl2XMOY1E4oWNE8mZyDqdW42kWIpEqJFzZCfy8ZfNR5iWJo4U4'
}

In [12]:
# replae with your EMR cluster ID
CLUSTER_ID = 'j-2ELHWKR7TCCMD'

def submit_job(app_name, pyfile_uri):
    emr = session.client('emr')
    emr.add_job_flow_steps(JobFlowId=CLUSTER_ID, Steps=[{
        'Name': app_name,
        'ActionOnFailure': 'CANCEL_AND_WAIT',
        'HadoopJarStep': {
            'Args': ['spark-submit',
                     '--master', 'yarn',
                     '--deploy-mode', 'cluster',
                     pyfile_uri],
            'Jar': 'command-runner.jar'
        }}])

In [ ]:
!pip install boto3
import boto3, json

session = boto3.session.Session(**credentials)
s3 = session.client('s3')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.6/135.6 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.7/10.7 MB 62.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.6/79.6 kB 9.7 MB/s eta 0:00:00


In [13]:
# EMR
%%file testing.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark import SparkContext, SparkConf
import json
from operator import add
#import other things as required

#create spark context. This is very important. Do this similarly for the other parts
# Note to read a file directly from s3 into an rdd you may have to do something like this

def checkjson(entry):
  try:
    json.loads(entry)
    #if load succeeded. We use correct as the key
    return "correct", 1
  except:
    #there was an error in loading. We use incorrect as the key
    return "incorrect", 1


if __name__ == '__main__':
  # replace this line with the s3 pass when testing over EMR
  conf = SparkConf().setAppName('testing').set('spark.hadoop.validateOutputSpecs', False)
  sc = SparkContext(conf=conf).getOrCreate()

  try:
    #@todo: fix the path as required
    tweets=sc.textFile('s3://cs4266-finalproject/nashville-tweets-2019-01-28')
    # review the page rank example for how to use the map operation
    # review word count for reduce and add
    # see how we use map to parse each row
    counts = tweets.map(lambda tweet: checkjson(tweet)).reduceByKey(add)
    print(counts.take(2))

    # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
    output = counts.sortBy(lambda a: a[1], ascending=False)

    # @todo: the s3 version will have to save it to correct s3 path
    output.repartition(1).saveAsTextFile("s3://cs4266-finalproject-spark/emr-logs/testing.out")
    

  finally:
    # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
    #finally is used to make sure the context is stopped even with errors
    sc.stop()
  

 
  
  pass

Writing testing.py


In [14]:
s3.upload_file(Filename='testing.py', Bucket='cs4266-finalproject-spark', Key='emr-logs/testing.py')

In [15]:
# submit spark job to emr. Make all the necessary changes to the path
submit_job(app_name='testing', pyfile_uri='s3://cs4266-finalproject-spark/emr-logs/testing.py')

In [16]:
def test1(lines):
  if '6294' in lines[0] and 'correct' in lines[0]:
    print("passed")
  else:
    print("failed")

# test local execution results
with open('local_1_count.out/part-00000') as f:
  lines = f.readlines()
  print(lines)
  test1(lines)

["('correct', 6294)\n"]
passed


In [22]:
# test EMR execution results. Once again, make sure that S3 paths are consistent.
output_key = "emr-logs/testing.out/part-00000"
lines = s3.get_object(Bucket='cs4266-finalproject-spark', Key=output_key)['Body'].read().decode().splitlines()
test1(lines)

passed


# We have verified that the old homework question works with our EMR and S3 buckets...

### Now, let's move on to trying to read in a basic parquet file

In [44]:
# EMR
%%file testing2.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark.sql import SparkSession
from operator import add

#create spark session
spark = SparkSession.builder.appName('testing2').getOrCreate()

try:
  # read in parquet file
  # part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet
  tweets_df = spark.read.parquet('s3://cs4266-finalproject/weather_tn.parquet/year=2021/month=1/part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet')
  tweets_df.createOrReplaceTempView("tweets")

  # run SQL query
  counts = spark.sql('SELECT COUNT(*) FROM tweets')
  print(counts.show())

  # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
  output = counts.sort("count", ascending=False)

  # @todo: the s3 version will have to save it to correct s3 path
  output.repartition(1).write.mode('overwrite').csv("s3://cs4266-finalproject-spark/emr-logs/testing2.out")
    

finally:
  # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
  #finally is used to make sure the context is stopped even with errors
  spark.stop()
pass


Overwriting testing2.py


In [45]:
s3.upload_file(Filename='testing2.py', Bucket='cs4266-finalproject-spark', Key='emr-logs/testing2.py')

In [46]:
# submit spark job to emr. Make all the necessary changes to the path
submit_job(app_name='testing2', pyfile_uri='s3://cs4266-finalproject-spark/emr-logs/testing2.py')

In [33]:
output_key = "emr-logs/testing2.out/part-00000-c588163a-1b46-4c13-a474-0b931eba155a-c000.csv"
lines = s3.get_object(Bucket='cs4266-finalproject-spark', Key=output_key)['Body'].read().decode().splitlines()

In [34]:
lines

['29765']

WE GOT 29765 ROWS FOR ONE PARQUET

# TRYING A BRAVE NEW ATTEMPT TO LOAD IT IN LOCALLY TO TEST SPARK

In [62]:
full_path = 'weather_tn.parquet/year=2021/month=1/part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet'

In [63]:
s3.download_file('cs4266-finalproject', full_path, 'weather.parquet')

In [61]:
s3.download_file('cs4266-finalproject', 'nfd_incidents_xd_seg.parquet', 'incidents.parquet')

In [68]:
%%file localtest2.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark.sql import SparkSession
from operator import add

#create spark session
spark = SparkSession.builder.appName('localtest2').getOrCreate()

try:
  # read in parquet file
  # part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet
  tweets_df = spark.read.parquet('weather.parquet')
  tweets_df.createOrReplaceTempView("weather")

  # run SQL query
  counts = spark.sql('SELECT COUNT(*) FROM weather')
  print(counts.show())

  # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
  output = counts#.sort("count", ascending=False)

  # @todo: the s3 version will have to save it to correct s3 path
  output.repartition(1).write.mode('overwrite').csv("localtest2.out")
    

finally:
  # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
  #finally is used to make sure the context is stopped even with errors
  spark.stop()
pass

Overwriting localtest2.py


In [69]:
!spark-submit --packages org.apache.spark:spark-streaming-kafka-0-8_2.11:2.4.7 localtest2.py

:: loading settings :: url = jar:file:/content/spark-3.2.4-bin-hadoop3.2/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /root/.ivy2/cache
The jars for the packages stored in: /root/.ivy2/jars
org.apache.spark#spark-streaming-kafka-0-8_2.11 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-9ad98076-9171-4319-8414-81f5ed5f0aff;1.0
	confs: [default]
	found org.apache.spark#spark-streaming-kafka-0-8_2.11;2.4.7 in central
	found org.apache.kafka#kafka_2.11;0.8.2.1 in central
	found org.scala-lang.modules#scala-xml_2.11;1.0.2 in central
	found com.yammer.metrics#metrics-core;2.2.0 in central
	found org.slf4j#slf4j-api;1.7.16 in central
	found org.scala-lang.modules#scala-parser-combinators_2.11;1.1.0 in central
	found com.101tec#zkclient;0.3 in central
	found log4j#log4j;1.2.17 in central
	found org.apache.kafka#kafka-clients;0.8.2.1 in central
	found net.jpountz.lz4#lz4;1.2.0 in central
	found org.xerial.snapp

In [70]:
with open('localtest2.out/part-00000-d87ee3a4-9e6a-4b96-98ee-f4fc22acc9cc-c000.csv') as f:
  lines = f.readlines()
  print(lines)

['144336\n']


# LOCALTEST2 WORKED... LET'S TRY TO SEND IT TO EMR!

In [77]:
%%file emrtest3.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark.sql import SparkSession
from operator import add

#create spark session
spark = SparkSession.builder.appName('emrtest3').getOrCreate()

try:
  # read in parquet file
  # part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet
  tweets_df = spark.read.parquet('s3://weather_tn.parquet:*/*/*.parquet')
  tweets_df.createOrReplaceTempView("weather")

  # run SQL query
  counts = spark.sql('SELECT COUNT(*) FROM weather')
  print(counts.show())

  # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
  output = counts#.sort("count", ascending=False)

  # @todo: the s3 version will have to save it to correct s3 path
  output.repartition(1).write.mode('overwrite').csv("emrtest3.out")
    

finally:
  # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
  #finally is used to make sure the context is stopped even with errors
  spark.stop()
pass

Overwriting emrtest3.py


In [75]:
s3.upload_file(Filename='emrtest3.py', Bucket='cs4266-finalproject-spark', Key='emr-logs/emrtest3.py')

In [76]:
# submit spark job to emr. Make all the necessary changes to the path
submit_job(app_name='emrtest3', pyfile_uri='s3://cs4266-finalproject-spark/emr-logs/emrtest3.py')

# That failed miserably... let's try some other approaches

In [81]:
%%file emrtest3.py
'''
TODO:
Count the number of tweets.
Parse tweets with json.loads -- note how the tweets are huge JSON blobs.
Ignore tweets that error on load.
'''

from pyspark.sql import SparkSession
from operator import add

#create spark session
spark = SparkSession.builder.appName('emrtest3').getOrCreate()

try:
  # read in parquet file
  # part-00030-0c8142a9-ffb0-4186-8d5f-060d4d2afe7d.c000.snappy.parquet
  df = spark.read.option("mergeSchema", "true").parquet("s3://weather_tn.parquet")
  df.createOrReplaceTempView("weather")

  # run SQL query
  results = spark.sql('SELECT COUNT(*) FROM weather WHERE year = "2021" AND month = "1"')

  # @todo: create an output rdd that uses reduce by key and add operator to add up the correct entries and incorrect entries
  output = results

  # @todo: the s3 version will have to save it to correct s3 path
  output.repartition(1).write.mode('overwrite').csv("emrtest3.out")


  """
sales_data = spark.read.format('parquet').load('s3://bucket/sales_data')

# Create a temporary view
sales_data.createOrReplaceTempView('sales')

# Run a SQL query on the partitioned data
results = spark.sql('SELECT SUM(sales_amount) FROM sales WHERE year = 2021 AND month = 01')
  """
    

finally:
  # very important: stop the context. Otherwise you may get an error that context is still alive. if you are on colab just restart the runtime if you face problem
  #finally is used to make sure the context is stopped even with errors
  spark.stop()
pass

Overwriting emrtest3.py


In [82]:
s3.upload_file(Filename='emrtest3.py', Bucket='cs4266-finalproject-spark', Key='emr-logs/emrtest3.py')

In [83]:
# submit spark job to emr. Make all the necessary changes to the path
submit_job(app_name='emrtest3', pyfile_uri='s3://cs4266-finalproject-spark/emr-logs/emrtest3.py')